JDK HTTP adapter fails when Vertx classes not on classpath - NoClassDefFoundError

## Problem Description

When using the JDK HTTP adapter (`kiota-http-jdk`) **without** Vertx dependencies on the classpath, the SDK fails with `NoClassDefFoundError` even though the JDK adapter is properly configured.

### Error

```
java.lang.NoClassDefFoundError: io/vertx/core/net/TrustOptions
    at io.apicurio.registry.client.RegistryClientFactory.create(RegistryClientFactory.java:24)
    at io.apicurio.registry.resolver.client.RegistryClientFacadeFactory.create_v3(RegistryClientFacadeFactory.java:167)
    ...
Caused by: java.lang.ClassNotFoundException: io.vertx.core.net.TrustOptions
```

### Configuration

```java
Map<String, Object> config = Map.of(
    "apicurio.registry.url", "http://localhost:8080/apis/registry/v3",
    SchemaResolverConfig.HTTP_ADAPTER, "JDK"
);
```

### Dependencies

```xml
<dependency>
    <groupId>io.apicurio</groupId>
    <artifactId>apicurio-registry-schema-resolver</artifactId>
    <version>3.1.7</version>
    <exclusions>
        <exclusion>
            <groupId>io.kiota</groupId>
            <artifactId>kiota-http-vertx</artifactId>
        </exclusion>
    </exclusions>
</dependency>
<dependency>
    <groupId>io.kiota</groupId>
    <artifactId>kiota-http-jdk</artifactId>
    <version>0.0.31</version>
</dependency>
```

## Root Cause Analysis

The issue is in `RegistryClientRequestAdapterFactory.java` which has **compile-time imports** of Vertx classes at the top of the file:

```java
// Lines 8-19
import io.kiota.http.vertx.VertXRequestAdapter;
import io.vertx.core.Vertx;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.http.HttpClosedException;
import io.vertx.core.net.JksOptions;
import io.vertx.core.net.PemKeyCertOptions;
import io.vertx.core.net.PemTrustOptions;
import io.vertx.core.net.PfxOptions;
import io.vertx.core.net.ProxyOptions;
import io.vertx.ext.web.client.WebClient;
import io.vertx.ext.web.client.WebClientOptions;
```

Similarly, `RegistryClientFacadeFactory.java` (in schema-resolver) imports:

```java
import io.vertx.core.Vertx;
```

When the JVM loads these classes, it attempts to resolve **all** import statements immediately, before any code executes. If Vertx classes are not on the classpath, the class loading fails with `NoClassDefFoundError`.

The `AdapterDetector` class correctly detects which adapter is available at runtime, but this detection happens **after** class loading, so it's too late.

## Proposed Solution

Refactor the Vertx-specific code into separate classes that are only loaded dynamically when the Vertx adapter is actually selected. This follows the "lazy loading" pattern commonly used to avoid compile-time dependencies on optional libraries.

### Option 1: Separate Adapter Factory Classes (Recommended)

Create adapter-specific factory classes:

**1. Create `VertxAdapterFactory.java`:**

```java
package io.apicurio.registry.client.common;

import io.kiota.http.vertx.VertXRequestAdapter;
import io.vertx.core.Vertx;
import io.vertx.core.net.*;
import io.vertx.ext.web.client.*;
// ... all Vertx imports

/**
 * Factory for creating Vertx-based request adapters.
 * This class is only loaded when Vertx adapter is selected.
 */
public class VertxAdapterFactory {
    
    public static RequestAdapter createAdapter(RegistryClientOptions options) {
        // Move all Vertx-specific code from RegistryClientRequestAdapterFactory here
    }
    
    public static Vertx getVertx(RegistryClientOptions options) {
        // Move Vertx instance resolution here
    }
    
    // ... other Vertx-specific methods
}
```

**2. Create `JdkAdapterFactory.java`:**

```java
package io.apicurio.registry.client.common;

import io.kiota.http.jdk.JDKRequestAdapter;
import java.net.http.HttpClient;
// ... only JDK imports

/**
 * Factory for creating JDK-based request adapters.
 * This class has no Vertx dependencies.
 */
public class JdkAdapterFactory {
    
    public static RequestAdapter createAdapter(RegistryClientOptions options) {
        // Move all JDK-specific code here
    }
}
```

**3. Update `RegistryClientRequestAdapterFactory.java`:**

```java
package io.apicurio.registry.client.common;

import com.microsoft.kiota.RequestAdapter;
// NO Vertx imports here!

public class RegistryClientRequestAdapterFactory {

    public static RequestAdapter createRequestAdapter(RegistryClientOptions options, Version version) {
        // ... validation code ...
        
        HttpAdapterType adapterType = AdapterDetector.resolveAdapterType(options.getHttpAdapterType());
        
        RequestAdapter adapter;
        switch (adapterType) {
            case VERTX:
                // Load Vertx factory dynamically - only fails if Vertx not available AND selected
                adapter = VertxAdapterFactory.createAdapter(options);
                break;
            case JDK:
                adapter = JdkAdapterFactory.createAdapter(options);
                break;
            default:
                throw new IllegalArgumentException("Unknown adapter type: " + adapterType);
        }
        
        // ... rest of the code ...
    }
}
```

**4. Update `RegistryClientFacadeFactory.java` (schema-resolver):**

Remove the `io.vertx.core.Vertx` import and use `Object` or a wrapper:

```java
// Before
import io.vertx.core.Vertx;
private static RegistryClientFacade create_v3(SchemaResolverConfig config, Vertx vertx) { ... }

// After - no Vertx import needed
private static RegistryClientFacade create_v3(SchemaResolverConfig config, Object vertx) { ... }
```

Or better, encapsulate the Vertx instance in `RegistryClientOptions` and don't pass it separately.

### Option 2: Reflection-based Loading

Use reflection to load Vertx classes only when needed:

```java
private static RequestAdapter createVertxAdapter(RegistryClientOptions options) {
    try {
        Class<?> factoryClass = Class.forName("io.apicurio.registry.client.common.VertxAdapterFactory");
        Method createMethod = factoryClass.getMethod("createAdapter", RegistryClientOptions.class);
        return (RequestAdapter) createMethod.invoke(null, options);
    } catch (ClassNotFoundException e) {
        throw new IllegalStateException("Vertx adapter factory not available", e);
    } catch (Exception e) {
        throw new RuntimeException("Failed to create Vertx adapter", e);
    }
}
```

## Impact

This issue affects any user who wants to:
- Use the JDK HTTP adapter exclusively
- Exclude Vertx dependencies to reduce artifact size (~15 MB)
- Use JDK HttpClient in environments where Vertx is problematic (GraalVM native, certain classloader scenarios)

## Affected Versions

- 3.1.7 (confirmed)
- Likely all 3.x versions with JDK adapter support

## Workaround

Currently, the only workaround is to keep Vertx dependencies on the classpath even when using the JDK adapter, which defeats the purpose of having an optional JDK adapter.

--- a/java-sdk-common/src/main/java/io/apicurio/registry/client/common/RegistryClientRequestAdapterFactory.java
+++ b/java-sdk-common/src/main/java/io/apicurio/registry/client/common/RegistryClientRequestAdapterFactory.java
@@ -3,24 +3,11 @@ package io.apicurio.registry.client.common;
 import com.microsoft.kiota.RequestAdapter;
 import com.microsoft.kiota.RequestInformation;
 import io.apicurio.registry.client.common.auth.JdkAuthFactory;
-import io.apicurio.registry.client.common.auth.VertXAuthFactory;
 import io.apicurio.registry.client.common.ssl.JdkSslContextFactory;
 import io.kiota.http.jdk.JDKRequestAdapter;
-import io.kiota.http.vertx.VertXRequestAdapter;
-import io.vertx.core.Vertx;
-import io.vertx.core.buffer.Buffer;
-import io.vertx.core.http.HttpClosedException;
-import io.vertx.core.net.JksOptions;
-import io.vertx.core.net.PemKeyCertOptions;
-import io.vertx.core.net.PemTrustOptions;
-import io.vertx.core.net.PfxOptions;
-import io.vertx.core.net.ProxyOptions;
-import io.vertx.ext.web.client.WebClient;
-import io.vertx.ext.web.client.WebClientOptions;
 
 import javax.net.ssl.SSLContext;
 import javax.net.ssl.SSLParameters;
-import java.lang.annotation.Annotation;
 import java.lang.reflect.InvocationHandler;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
@@ -80,7 +67,8 @@ public class RegistryClientRequestAdapterFactory {
         RequestAdapter adapter;
         switch (adapterType) {
             case VERTX:
-                adapter = createVertxAdapter(options);
+                // Delegate to VertxAdapterFactory to avoid loading Vertx classes when JDK adapter is used
+                adapter = VertxAdapterFactory.createAdapter(options);
                 break;
             case JDK:
                 adapter = createJdkAdapter(options);
@@ -97,182 +85,6 @@ public class RegistryClientRequestAdapterFactory {
         }
 
         return adapter;
-    }
-
-    // ==================== Vert.x Adapter Implementation ====================
-
-    private static RequestAdapter createVertxAdapter(RegistryClientOptions options) {
-        Vertx vertxToUse = getVertx(options);
-        WebClientOptions webClientOptions = buildWebClientOptions(options);
-
-        switch (options.getAuthType()) {
-            case ANONYMOUS:
-                return createVertxAnonymous(vertxToUse, webClientOptions);
-            case BASIC:
-                return createVertxBasicAuth(options.getUsername(), options.getPassword(), vertxToUse, webClientOptions);
-            case OAUTH2:
-                return createVertxOAuth2(options.getTokenEndpoint(), options.getClientId(),
-                        options.getClientSecret(), options.getScope(), vertxToUse, webClientOptions);
-            case CUSTOM_WEBCLIENT:
-                return createVertxCustomWebClient(options.getWebClient());
-            default:
-                throw new IllegalArgumentException("Unsupported authentication type: " + options.getAuthType());
-        }
-    }
-
-    private static Vertx getVertxFromCDI(String CDIClassName, String InstanceClassName) {
-        try {
-            var CDIClass = Class.forName(CDIClassName);
-            var instanceClass = Class.forName(InstanceClassName);
-            var CDI = CDIClass.getMethod("current").invoke(null);
-            var vertxInstance = CDIClass.getMethod("select", Class.class, Annotation[].class).invoke(CDI, Vertx.class, new Annotation[]{});
-            return (Vertx) instanceClass.getMethod("get").invoke(vertxInstance);
-        } catch (Throwable t) {
-            log.log(Level.FINE, "Attempt to retrieve a Vertx instance from CDI failed: "
-                    + t.getClass().getCanonicalName() + ": " + t.getMessage());
-            return null;
-        }
-    }
-
-    private static Vertx getVertx(RegistryClientOptions options) {
-        if (options.getVertx() != null) {
-            return options.getVertx();
-        }
-
-        var vertx = getVertxFromCDI("jakarta.enterprise.inject.spi.CDI", "jakarta.enterprise.inject.Instance");
-        if (vertx == null) {
-            vertx = getVertxFromCDI("javax.enterprise.inject.spi.CDI", "javax.enterprise.inject.Instance");
-        }
-        if (vertx != null) {
-            log.log(Level.FINE, "Successfully retrieved a Vertx instance from CDI.");
-            return vertx;
-        }
-
-        return DefaultVertxInstance.get();
-    }
-
-    private static RequestAdapter createVertxAnonymous(Vertx vertx, WebClientOptions webClientOptions) {
-        WebClient webClient = webClientOptions == null ? WebClient.create(vertx) : WebClient.create(vertx, webClientOptions);
-        return new VertXRequestAdapter(webClient);
-    }
-
-    private static RequestAdapter createVertxBasicAuth(String username, String password, Vertx vertx, WebClientOptions webClientOptions) {
-        WebClient webClient = VertXAuthFactory.buildSimpleAuthWebClient(vertx, webClientOptions, username, password);
-        return new VertXRequestAdapter(webClient);
-    }
-
-    private static RequestAdapter createVertxOAuth2(String tokenEndpoint,
-                                                     String clientId, String clientSecret, String scope, Vertx vertx, WebClientOptions webClientOptions) {
-        WebClient webClient = VertXAuthFactory.buildOIDCWebClient(vertx, webClientOptions, tokenEndpoint, clientId, clientSecret, scope);
-        return new VertXRequestAdapter(webClient);
-    }
-
-    private static RequestAdapter createVertxCustomWebClient(WebClient webClient) {
-        if (webClient == null) {
-            throw new IllegalArgumentException("WebClient cannot be null");
-        }
-        return new VertXRequestAdapter(webClient);
-    }
-
-    private static WebClientOptions buildWebClientOptions(RegistryClientOptions options) {
-        boolean hasSslConfig = options.getTrustStoreType() != RegistryClientOptions.TrustStoreType.NONE
-                || options.getKeyStoreType() != RegistryClientOptions.KeyStoreType.NONE
-                || options.isTrustAll()
-                || !options.isVerifyHost();
-
-        boolean hasProxyConfig = options.getProxyHost() != null;
-
-        if (!hasSslConfig && !hasProxyConfig) {
-            return null;
-        }
-
-        WebClientOptions webClientOptions = new WebClientOptions();
-
-        if (hasSslConfig) {
-            webClientOptions.setSsl(true);
-        }
-
-        if (options.isTrustAll()) {
-            webClientOptions.setTrustAll(true);
-        }
-
-        webClientOptions.setVerifyHost(options.isVerifyHost());
-
-        switch (options.getTrustStoreType()) {
-            case JKS:
-                JksOptions jksOptions = new JksOptions()
-                        .setPath(options.getTrustStorePath())
-                        .setPassword(options.getTrustStorePassword());
-                webClientOptions.setTrustOptions(jksOptions);
-                break;
-            case PKCS12:
-                PfxOptions pfxOptions = new PfxOptions()
-                        .setPath(options.getTrustStorePath())
-                        .setPassword(options.getTrustStorePassword());
-                webClientOptions.setTrustOptions(pfxOptions);
-                break;
-            case PEM:
-                PemTrustOptions pemOptions = new PemTrustOptions();
-                if (options.getPemCertContent() != null) {
-                    Buffer certBuffer = Buffer.buffer(options.getPemCertContent());
-                    pemOptions.addCertValue(certBuffer);
-                } else if (options.getPemCertPaths() != null) {
-                    for (String certPath : options.getPemCertPaths()) {
-                        pemOptions.addCertPath(certPath);
-                    }
-                }
-                webClientOptions.setTrustOptions(pemOptions);
-                break;
-            case NONE:
-                break;
-        }
-
-        switch (options.getKeyStoreType()) {
-            case JKS:
-                JksOptions jksKeyStoreOptions = new JksOptions()
-                        .setPath(options.getKeyStorePath())
-                        .setPassword(options.getKeyStorePassword());
-                webClientOptions.setKeyCertOptions(jksKeyStoreOptions);
-                break;
-            case PKCS12:
-                PfxOptions pfxKeyStoreOptions = new PfxOptions()
-                        .setPath(options.getKeyStorePath())
-                        .setPassword(options.getKeyStorePassword());
-                webClientOptions.setKeyCertOptions(pfxKeyStoreOptions);
-                break;
-            case PEM:
-                PemKeyCertOptions pemKeyCertOptions = new PemKeyCertOptions();
-                if (options.getPemClientCertContent() != null && options.getPemClientKeyContent() != null) {
-                    Buffer certBuffer = Buffer.buffer(options.getPemClientCertContent());
-                    Buffer keyBuffer = Buffer.buffer(options.getPemClientKeyContent());
-                    pemKeyCertOptions.addCertValue(certBuffer);
-                    pemKeyCertOptions.addKeyValue(keyBuffer);
-                } else if (options.getPemClientCertPath() != null && options.getPemClientKeyPath() != null) {
-                    pemKeyCertOptions.addCertPath(options.getPemClientCertPath());
-                    pemKeyCertOptions.addKeyPath(options.getPemClientKeyPath());
-                }
-                webClientOptions.setKeyCertOptions(pemKeyCertOptions);
-                break;
-            case NONE:
-                break;
-        }
-
-        if (hasProxyConfig) {
-            ProxyOptions proxyOptions = new ProxyOptions()
-                    .setHost(options.getProxyHost())
-                    .setPort(options.getProxyPort());
-
-            if (options.getProxyUsername() != null && !options.getProxyUsername().isEmpty()) {
-                proxyOptions.setUsername(options.getProxyUsername());
-                if (options.getProxyPassword() != null) {
-                    proxyOptions.setPassword(options.getProxyPassword());
-                }
-            }
-
-            webClientOptions.setProxyOptions(proxyOptions);
-        }
-
-        return webClientOptions;
     }
 
     // ==================== JDK Adapter Implementation ====================
@@ -447,8 +259,8 @@ public class RegistryClientRequestAdapterFactory {
         }
 
         private boolean isRetryable(Throwable cause) {
-            // Vert.x specific retryable exception
-            if (cause instanceof HttpClosedException) {
+            // Vert.x specific retryable exception - check by class name to avoid compile-time Vertx dependency
+            if ("io.vertx.core.http.HttpClosedException".equals(cause.getClass().getName())) {
                 return true;
             }
             // JDK specific retryable exceptions

--- a/java-sdk-common/src/main/java/io/apicurio/registry/client/common/VertxAdapterFactory.java
+++ b/java-sdk-common/src/main/java/io/apicurio/registry/client/common/VertxAdapterFactory.java
@@ -1,0 +1,218 @@
+package io.apicurio.registry.client.common;
+
+import com.microsoft.kiota.RequestAdapter;
+import io.apicurio.registry.client.common.auth.VertXAuthFactory;
+import io.kiota.http.vertx.VertXRequestAdapter;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.net.JksOptions;
+import io.vertx.core.net.PemKeyCertOptions;
+import io.vertx.core.net.PemTrustOptions;
+import io.vertx.core.net.PfxOptions;
+import io.vertx.core.net.ProxyOptions;
+import io.vertx.ext.web.client.WebClient;
+import io.vertx.ext.web.client.WebClientOptions;
+
+import java.lang.annotation.Annotation;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Factory class for creating Vert.x-based RequestAdapter instances.
+ *
+ * <p>This class is isolated from the main RegistryClientRequestAdapterFactory to ensure
+ * that Vert.x classes are only loaded when the Vert.x adapter is explicitly selected.
+ * This prevents NoClassDefFoundError when using the JDK adapter without Vert.x on the classpath.</p>
+ */
+public final class VertxAdapterFactory {
+
+    private static final Logger log = Logger.getLogger(VertxAdapterFactory.class.getName());
+
+    private VertxAdapterFactory() {
+        // Prevent instantiation
+    }
+
+    /**
+     * Creates a Vert.x-based RequestAdapter configured with authentication, SSL/TLS, and other settings
+     * from the provided options.
+     *
+     * @param options the configuration options
+     * @return a fully configured Vert.x RequestAdapter
+     * @throws IllegalArgumentException if options are invalid or authentication type is unsupported
+     */
+    public static RequestAdapter createAdapter(RegistryClientOptions options) {
+        Vertx vertxToUse = getVertx(options);
+        WebClientOptions webClientOptions = buildWebClientOptions(options);
+
+        switch (options.getAuthType()) {
+            case ANONYMOUS:
+                return createVertxAnonymous(vertxToUse, webClientOptions);
+            case BASIC:
+                return createVertxBasicAuth(options.getUsername(), options.getPassword(), vertxToUse, webClientOptions);
+            case OAUTH2:
+                return createVertxOAuth2(options.getTokenEndpoint(), options.getClientId(),
+                        options.getClientSecret(), options.getScope(), vertxToUse, webClientOptions);
+            case CUSTOM_WEBCLIENT:
+                return createVertxCustomWebClient(options);
+            default:
+                throw new IllegalArgumentException("Unsupported authentication type: " + options.getAuthType());
+        }
+    }
+
+    private static Vertx getVertxFromCDI(String CDIClassName, String InstanceClassName) {
+        try {
+            var CDIClass = Class.forName(CDIClassName);
+            var instanceClass = Class.forName(InstanceClassName);
+            var CDI = CDIClass.getMethod("current").invoke(null);
+            var vertxInstance = CDIClass.getMethod("select", Class.class, Annotation[].class).invoke(CDI, Vertx.class, new Annotation[]{});
+            return (Vertx) instanceClass.getMethod("get").invoke(vertxInstance);
+        } catch (Throwable t) {
+            log.log(Level.FINE, "Attempt to retrieve a Vertx instance from CDI failed: "
+                    + t.getClass().getCanonicalName() + ": " + t.getMessage());
+            return null;
+        }
+    }
+
+    private static Vertx getVertx(RegistryClientOptions options) {
+        Vertx vertx = options.getVertx();
+        if (vertx != null) {
+            return vertx;
+        }
+
+        vertx = getVertxFromCDI("jakarta.enterprise.inject.spi.CDI", "jakarta.enterprise.inject.Instance");
+        if (vertx == null) {
+            vertx = getVertxFromCDI("javax.enterprise.inject.spi.CDI", "javax.enterprise.inject.Instance");
+        }
+        if (vertx != null) {
+            log.log(Level.FINE, "Successfully retrieved a Vertx instance from CDI.");
+            return vertx;
+        }
+
+        return DefaultVertxInstance.get();
+    }
+
+    private static RequestAdapter createVertxAnonymous(Vertx vertx, WebClientOptions webClientOptions) {
+        WebClient webClient = webClientOptions == null ? WebClient.create(vertx) : WebClient.create(vertx, webClientOptions);
+        return new VertXRequestAdapter(webClient);
+    }
+
+    private static RequestAdapter createVertxBasicAuth(String username, String password, Vertx vertx, WebClientOptions webClientOptions) {
+        WebClient webClient = VertXAuthFactory.buildSimpleAuthWebClient(vertx, webClientOptions, username, password);
+        return new VertXRequestAdapter(webClient);
+    }
+
+    private static RequestAdapter createVertxOAuth2(String tokenEndpoint,
+                                                     String clientId, String clientSecret, String scope, Vertx vertx, WebClientOptions webClientOptions) {
+        WebClient webClient = VertXAuthFactory.buildOIDCWebClient(vertx, webClientOptions, tokenEndpoint, clientId, clientSecret, scope);
+        return new VertXRequestAdapter(webClient);
+    }
+
+    private static RequestAdapter createVertxCustomWebClient(RegistryClientOptions options) {
+        WebClient webClient = options.getWebClient();
+        if (webClient == null) {
+            throw new IllegalArgumentException("WebClient cannot be null");
+        }
+        return new VertXRequestAdapter(webClient);
+    }
+
+    private static WebClientOptions buildWebClientOptions(RegistryClientOptions options) {
+        boolean hasSslConfig = options.getTrustStoreType() != RegistryClientOptions.TrustStoreType.NONE
+                || options.getKeyStoreType() != RegistryClientOptions.KeyStoreType.NONE
+                || options.isTrustAll()
+                || !options.isVerifyHost();
+
+        boolean hasProxyConfig = options.getProxyHost() != null;
+
+        if (!hasSslConfig && !hasProxyConfig) {
+            return null;
+        }
+
+        WebClientOptions webClientOptions = new WebClientOptions();
+
+        if (hasSslConfig) {
+            webClientOptions.setSsl(true);
+        }
+
+        if (options.isTrustAll()) {
+            webClientOptions.setTrustAll(true);
+        }
+
+        webClientOptions.setVerifyHost(options.isVerifyHost());
+
+        switch (options.getTrustStoreType()) {
+            case JKS:
+                JksOptions jksOptions = new JksOptions()
+                        .setPath(options.getTrustStorePath())
+                        .setPassword(options.getTrustStorePassword());
+                webClientOptions.setTrustOptions(jksOptions);
+                break;
+            case PKCS12:
+                PfxOptions pfxOptions = new PfxOptions()
+                        .setPath(options.getTrustStorePath())
+                        .setPassword(options.getTrustStorePassword());
+                webClientOptions.setTrustOptions(pfxOptions);
+                break;
+            case PEM:
+                PemTrustOptions pemOptions = new PemTrustOptions();
+                if (options.getPemCertContent() != null) {
+                    Buffer certBuffer = Buffer.buffer(options.getPemCertContent());
+                    pemOptions.addCertValue(certBuffer);
+                } else if (options.getPemCertPaths() != null) {
+                    for (String certPath : options.getPemCertPaths()) {
+                        pemOptions.addCertPath(certPath);
+                    }
+                }
+                webClientOptions.setTrustOptions(pemOptions);
+                break;
+            case NONE:
+                break;
+        }
+
+        switch (options.getKeyStoreType()) {
+            case JKS:
+                JksOptions jksKeyStoreOptions = new JksOptions()
+                        .setPath(options.getKeyStorePath())
+                        .setPassword(options.getKeyStorePassword());
+                webClientOptions.setKeyCertOptions(jksKeyStoreOptions);
+                break;
+            case PKCS12:
+                PfxOptions pfxKeyStoreOptions = new PfxOptions()
+                        .setPath(options.getKeyStorePath())
+                        .setPassword(options.getKeyStorePassword());
+                webClientOptions.setKeyCertOptions(pfxKeyStoreOptions);
+                break;
+            case PEM:
+                PemKeyCertOptions pemKeyCertOptions = new PemKeyCertOptions();
+                if (options.getPemClientCertContent() != null && options.getPemClientKeyContent() != null) {
+                    Buffer certBuffer = Buffer.buffer(options.getPemClientCertContent());
+                    Buffer keyBuffer = Buffer.buffer(options.getPemClientKeyContent());
+                    pemKeyCertOptions.addCertValue(certBuffer);
+                    pemKeyCertOptions.addKeyValue(keyBuffer);
+                } else if (options.getPemClientCertPath() != null && options.getPemClientKeyPath() != null) {
+                    pemKeyCertOptions.addCertPath(options.getPemClientCertPath());
+                    pemKeyCertOptions.addKeyPath(options.getPemClientKeyPath());
+                }
+                webClientOptions.setKeyCertOptions(pemKeyCertOptions);
+                break;
+            case NONE:
+                break;
+        }
+
+        if (hasProxyConfig) {
+            ProxyOptions proxyOptions = new ProxyOptions()
+                    .setHost(options.getProxyHost())
+                    .setPort(options.getProxyPort());
+
+            if (options.getProxyUsername() != null && !options.getProxyUsername().isEmpty()) {
+                proxyOptions.setUsername(options.getProxyUsername());
+                if (options.getProxyPassword() != null) {
+                    proxyOptions.setPassword(options.getProxyPassword());
+                }
+            }
+
+            webClientOptions.setProxyOptions(proxyOptions);
+        }
+
+        return webClientOptions;
+    }
+}

